share/doc/revng/references/abi-definition.md

```diff
@@ -0,0 +1,213 @@
+
+> **_Note:_** If a return value doesn't fit into the specified registers or uses irregular padding, the registers are not used and the object is passed using the memory (stack, pointer-to-copy, etc.).
+
+## `MaximumGPRsPerScalarArgument` (`unsigned`)
+
+States the maximum number of GPRs available to pass a single scalar (`int`, `__int128`, pointer, etc.) argument, meaning that it can only be passed in the GPRs if `MaximumGPRsPerScalarArgument` is less than or equal to the number of the registers required to fit the object.
+
+> **_Note:_** If `MaximumGPRsPerScalarArgument` is equal to 0, it means that the ABI does not allow scalar arguments to use GPRs.
+
+> **_Note:_** If an argument doesn't fit into the specified registers or uses irregular padding, the registers are not used and the object is passed using the memory (stack, pointer-to-copy, etc.).
+
+## `MaximumGPRsPerScalarReturnValue` (`unsigned`)
+
+States the maximum number of GPRs available to return a single scalar (`int`, `__int128`, pointer, etc.) value, meaning that it can only be returned in the GPRs if `MaximumGPRsPerScalarReturnValue` is less than or equal to the number of the registers required to fit the object including padding.
+
+> **_Note:_** If `MaximumGPRsPerScalarReturnValue` is equal to 0, it means that the ABI does not allow scalar return values to use GPRs.
+
+> **_Note:_** If a return value doesn't fit into the specified registers or uses irregular padding, the registers are not used and the object is passed using the memory (stack, pointer-to-copy, etc.).
+
+## `GeneralPurposeArgumentRegisters` (`list of registers`)
+
+Specifies the list of general purpose registers allowed to be used for passing arguments and the order they are to be used in.
+
+## `GeneralPurposeReturnValueRegisters` (`list of registers`)
+
+Specifies the list of general purpose registers allowed to be used for returning values and the order they are to be used in.
+
+## `VectorArgumentRegisters` (`list of registers`)
+
+Specifies the list of vector registers allowed to be used for passing arguments and the order they are to be used in.
+
+## `VectorReturnValueRegisters` (`list of registers`)
+
+Specifies the list of vector registers allowed to be used for returning values and the order they are to be used in.
+
+## `CalleeSavedRegisters` (`list of registers`)
+
+Specifies the list of registers for which the ABI requires the callee to preserve the value, meaning that when the callee returns, the value of those registers must be the same as it was when the function was called.
+
+## `ReturnValueLocationRegister` (`register`)
+
+Specifies a register to be used for returning (or even passing, depending on ABI) the pointer to the memory used for returning copies of big aggregate objects.
+
+Must be set to `model::Register::Invalid` for ABIs that do not support returning values by 'pointer-to-copy' or use stack for it.
```

Comment:
1. Use `"` for `pointer-to-copy`.
2. I don't get if, in SystemV x86-64, this is `rdi` or `rax`. The "for returning (or even passing, ...)" part confuses me.
   If it's `rdi` I wouldn't say it's "the register to *return* the pointer to the memory", it's the register used to *pass* it.
3. I wonder if the situation with "return value locations" could be better represented with an `enum` instead of using multiple `bool`s.

Reply:
1. 👌 
2. Looks like a leftover from an old design, my bad. Yeah, it's only about passing the pointer in. Returning (when it happens) uses the first GPR return value register from the usual list.
3. It's only two bools though, so no idea if enum is worth the overhead. Don't forget that this first one is a register - and enum is not going to help us get rid of it.

Reply:
2. OK, make sure we clarify that and make the SystemV x86-64 example.
3. Ah, right. Grouping them together and (if it makes sense) cross-referencing each other will probably make the situation clearer.

Reply:
2.
> Specifies which register is used for passing the pointer to the memory used for returning big aggregate objects.

> For example, for `SystemV_x86_64` ABI it is set to `rdi` (the first argument register), but for `arm64` ABI it is set to a dedicated `x8` register (hence the need for a separate option).

> It must be set to `model::Register::Invalid` for ABIs that do not support returning values by "pointer-to-copy" or use stack for it.

Reply:
Perfect.

Reply:
Done.